src/bindings/jna/libelektra/src/test/java/org/libelektra/KeySetTest.java

```diff
@@ -814,4 +809,500 @@ public void test_keySetTailSetShouldPass() {
+    assertEquals(key3, keySet.floor(key3));
+    assertEquals(key4, keySet.floor(key4));
+    assertEquals(key5, keySet.floor(key5));
+    assertNull(keySet.floor(key6));
+  }
+
+  @Test
+  public void test_keySetCeiling_shouldPass() {
+    var keySet = KeySet.create(5, key2, key3, key4, key5);
+
+    assertNull(keySet.ceiling(key));
+    assertEquals(key2, keySet.ceiling(key2));
+    assertEquals(key3, keySet.ceiling(key3));
+    assertEquals(key4, keySet.ceiling(key4));
+    assertEquals(key5, keySet.ceiling(key5));
+    assertNull(keySet.ceiling(key6));
+  }
+
+  @Test
+  public void test_keySetHigher_shouldPass() {
+    var keySet = KeySet.create(5, key2, key3, key4, key5);
+
+    assertNull(keySet.higher(key));
+    assertEquals(key3, keySet.higher(key2));
+    assertEquals(key4, keySet.higher(key3));
+    assertEquals(key5, keySet.higher(key4));
+    assertNull(keySet.higher(key5));
+    assertNull(keySet.higher(key6));
+  }
+
+  @Test
+  public void test_keySetPollFirst_shouldPass() {
+    var keySet = KeySet.create(2, key, key5);
+
+    assertEquals(2, keySet.size());
+    assertTrue(keySet.lookup(key).isPresent());
+    assertEquals(key, keySet.first());
+    var oPolledKey = keySet.pollFirst();
+    assertNotNull(oPolledKey);
+    assertEquals(key.getName(), oPolledKey.getName());
+    assertEquals(1, keySet.size());
+    assertTrue(keySet.lookup(key).isEmpty());
+
+    assertEquals(key5, keySet.pollFirst());
+    assertNull(keySet.pollFirst());
+  }
+
+  @Test
+  public void test_keySetPollLast_shouldPass() {
+    var keySet = KeySet.create(5, key2, key3, key4, key5, key6);
+
+    assertEquals(5, keySet.size());
+    assertTrue(keySet.lookup(key6).isPresent());
+    assertEquals(key6, keySet.last());
+    var oPolledKey = keySet.pollLast();
+    assertNotNull(oPolledKey);
+    assertEquals(key6.getName(), oPolledKey.getName());
+    assertEquals(4, keySet.size());
+    assertTrue(keySet.lookup(key6).isEmpty());
+
+    assertNull(KeySet.create().pollLast());
+  }
+
+  @Test
+  public void test_keySetDescendingSet_shouldPass() {
+    var keySet = KeySet.create(3, key3, key4, key5).descendingSet();
+    // key5, key4, key3
+
+    assertEquals(key5, keySet.first());
+    assertEquals(key3, keySet.last());
+
+    assertNull(keySet.lower(key6));
+    assertNull(keySet.lower(key5));
+    assertEquals(key5, keySet.lower(key4));
+    assertEquals(key4, keySet.lower(key3));
+
+    assertNull(keySet.floor(key6));
+    assertEquals(key5, keySet.floor(key5));
+    assertEquals(key4, keySet.floor(key4));
+    assertEquals(key3, keySet.floor(key3));
+
+    assertNull(keySet.ceiling(key6));
+    assertEquals(key5, keySet.ceiling(key5));
+    assertEquals(key4, keySet.ceiling(key4));
+    assertEquals(key3, keySet.ceiling(key3));
+
+    assertEquals(key4, keySet.higher(key5));
+    assertEquals(key3, keySet.higher(key4));
+    assertNull(keySet.higher(key3));
+    assertNull(keySet.higher(key2));
+    assertNull(keySet.higher(key));
```

Comment:
Some of this assertions are also wrong just like for `KeySet` above (but now obviously the other way around). And again, you don't check with missing keys in the middle. An example here would be:

```java
assertEquals(key5, keySet.lower(Key.create("/key_test/4/key_name1")));
```